contracts/core/Registry.sol

```diff
@@ -1,47 +1,315 @@
-        address attestationAddr;
+        Metadata metadata;
+        address anchor;
+    }
+
+    /// ==========================
+    /// === Storage Variables ====
+    /// ==========================
+
+    /// @notice Identity.id -> Identity
+    mapping(bytes32 => Identity) public identitiesById;
+
+    /// @notice anchor -> Identity.id
+    mapping(address => bytes32) public anchorToIdentityId;
+
+    /// ======================
+    /// ======= Events =======
+    /// ======================
+
+    event IdentityCreated(
+        bytes32 indexed identityId,
+        uint nonce,
+        string name,
+        Metadata metadata,
+        address anchor
+    );
+    event IdentityNameUpdated(
+        bytes32 indexed identityId,
+        string name,
+        address anchor
+    );
+    event IdentityMetadataUpdated(
+        bytes32 indexed identityId,
+        Metadata metadata
+    );
+
+    /// ======================
+    /// ==== Constructor =====
+    /// ======================
+    constructor() {
+        // Empty constructor
+    }
+
+    /// ====================================
+    /// ==== External/Public Functions =====
+    /// ====================================
+
+    /// @notice Retrieve identity by identityId
+    /// @param identityId The identityId of the identity
+    function getIdentity(bytes32 identityId) public view returns (Identity memory) {
+        return identitiesById[identityId];
+    }
+
+    /// @notice Retrieve identity by anchor
+    /// @param anchor The anchor of the identity
+    function getIdentityByAnchor(address anchor) public view returns (Identity memory) {
+        bytes32 identityId = anchorToIdentityId[anchor];
+        return identitiesById[identityId];
+    }
+
+    /// @notice Creates a new identity
+    /// @dev This will also set the attestation address generated from msg.sender and name
+    /// @param _nonce Index of the identity
+    /// @param _name The name of the identity
+    /// @param _metadata The metadata of the identity
+    /// @param _owner The owner of the identity
+    /// @param _members The members of the identity
+    function createIdentity(
+        uint _nonce,
+        string memory _name,
+        Metadata memory _metadata,
+        address _owner,
+        address[] memory _members
+    ) external returns (bytes32) {
+        bytes32 identityId = _generateIdentityId(_nonce);
+
+        if (identitiesById[identityId].id != bytes32(0)) {
+            revert INDEX_NOT_AVAILABLE();
+        }
+
+        Identity memory identity = Identity(
+            identityId,
+            _nonce,
+            _name,
+            _metadata,
+            _generateAnchor(identityId, _name)
+        );
+
+        identitiesById[identityId] = identity;
+        anchorToIdentityId[identity.anchor] = identityId;
+
+        // generate roles
+        bytes32 ownerRole = _generateRole(identityId, RoleType.OWNER);
+        bytes32 memberRole = _generateRole(identityId, RoleType.MEMBER);
```

Comment:
we shouls think about storing the roles on the struct
so we don't need to generate them again on every use

Reply:
I quite like that. 
We have to generate them multiple times otherwise.

Reply:
We are going to move the `owner` to the `Identity` struct per our conversation with @zobront this morning.

```javascript
struct Identity {
    uint nonce;
    address owner;
    string name;
    Metadata metadata;
    address anchor;
}
```